File: i2pchat/groups/invite.py
from __future__ import annotations

import base64
import json
import secrets
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any

from i2pchat import crypto

from .models import normalize_member_id, utc_now

GROUP_INVITE_PREFIX = "__I2PCHAT_GROUP_INVITE__:"
# Protocol v4 (I2PChat 1.4.0): invites are Ed25519-signed by the inviter's
# handshake signing key. v1 (unsigned) invites are no longer accepted.
GROUP_INVITE_VERSION = 2
MAX_GROUP_INVITE_BYTES = 256 * 1024
_INVITE_SIG_DOMAIN = b"I2PCHAT-GROUP-INVITE-v2"
# Outer seal: random wrap key + SecretBox(JSON). The shareable string is
# base64url with no stable prefix, so title/members/destinations are not
# recognizable on sight. Anyone who has the full token can still decrypt it.
_INVITE_WRAP_KEY_SIZE = 32
_INVITE_WRAP_SALT = b"I2PCHAT-GROUP-INVITE-SEAL"
_INVITE_WRAP_INFO = b"I2PCHAT-GROUP-INVITE-SEAL|v3"
_INVITE_B64URL_ALPHABET = frozenset(
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789_-"
)
# wrap_key(32) + SecretBox overhead (nonce 24 + mac 16) + tiny JSON.
_MIN_SEALED_INVITE_BYTES = _INVITE_WRAP_KEY_SIZE + 24 + 16 + 16


def _to_iso8601(value: datetime) -> str:
    return value.astimezone(timezone.utc).isoformat()


def _parse_datetime(value: str) -> datetime:
    parsed = datetime.fromisoformat(str(value).replace("Z", "+00:00"))
    if parsed.tzinfo is None:
        return parsed.replace(tzinfo=timezone.utc)
    return parsed.astimezone(timezone.utc)


def _normalize_members(raw_members: Any) -> tuple[str, ...]:
    if not isinstance(raw_members, (list, tuple)):
        raise ValueError("Group invite members must be a list")
    members: list[str] = []
    seen: set[str] = set()
    for raw in raw_members:
        member_id = normalize_member_id(str(raw or ""))
        if not member_id or member_id in seen:
            continue
        seen.add(member_id)
        members.append(member_id)
    if not members:
        raise ValueError("Group invite must include at least one member")
    return tuple(members)


@dataclass(slots=True, frozen=True)
class GroupInvite:
    invite_id: str
    group_id: str
    members: tuple[str, ...]
    epoch: int
    inviter_id: str
    title: str | None = None
    created_at: datetime | None = None
    version: int = GROUP_INVITE_VERSION
    # v2: Ed25519 identity (handshake signing key) of the inviter and, once
    # encoded, the detached signature over the canonical invite payload.
    inviter_signing_pub: str = ""
    expires_at: datetime | None = None
    signature: str = ""

    def __post_init__(self) -> None:
        invite_id = (self.invite_id or "").strip()
        group_id = (self.group_id or "").strip()
        inviter_id = normalize_member_id(self.inviter_id)
        title = (self.title or "").strip() or None
        if not invite_id:
            raise ValueError("Group invite invite_id is required")
        if not group_id:
            raise ValueError("Group invite group_id is required")
        if not inviter_id:
            raise ValueError("Group invite inviter_id is required")
        members = _normalize_members(self.members)
        if inviter_id not in members:
            members = members + (inviter_id,)
        object.__setattr__(self, "invite_id", invite_id)
        object.__setattr__(self, "group_id", group_id)
        object.__setattr__(self, "inviter_id", inviter_id)
        object.__setattr__(self, "title", title)
        object.__setattr__(self, "members", members)
        object.__setattr__(self, "epoch", int(self.epoch))
        object.__setattr__(self, "version", int(self.version))
        object.__setattr__(
            self, "inviter_signing_pub", (self.inviter_signing_pub or "").strip().lower()
        )
        object.__setattr__(self, "signature", (self.signature or "").strip().lower())


def build_group_invite(
    *,
    group_id: str,
    members: list[str] | tuple[str, ...],
    epoch: int,
    inviter_id: str,
    title: str | None = None,
    invite_id: str | None = None,
    created_at: datetime | None = None,
    expires_at: datetime | None = None,
) -> GroupInvite:
    return GroupInvite(
        invite_id=(invite_id or secrets.token_hex(8)).strip(),
        group_id=group_id,
        members=tuple(members),
        epoch=int(epoch),
        inviter_id=inviter_id,
        title=title,
        created_at=created_at or utc_now(),
        expires_at=expires_at,
        version=GROUP_INVITE_VERSION,
    )


def _canonical_invite_bytes(
    *,
    version: int,
    invite_id: str,
    group_id: str,
    title: str | None,
    members: tuple[str, ...],
    epoch: int,
    inviter_id: str,
    inviter_signing_pub: str,
    created_at: str,
    expires_at: str | None,
) -> bytes:
    """Deterministic serialization signed by the inviter (excludes signature)."""
    payload = {
        "created_at": created_at,
        "epoch": int(epoch),
        "expires_at": expires_at,
        "group_id": group_id,
        "invite_id": invite_id,
        "inviter_id": inviter_id,
        "inviter_signing_pub": inviter_signing_pub,
        "members": list(members),
        "title": title,
        "v": int(version),
    }
    body = json.dumps(payload, sort_keys=True, separators=(",", ":"), ensure_ascii=True)
    return _INVITE_SIG_DOMAIN + b"|" + body.encode("utf-8")


def encode_group_invite(invite: GroupInvite, signing_seed: bytes) -> str:
    """Serialize and Ed25519-sign an invite with the inviter's signing seed.

    ``signing_seed`` is the inviter's handshake signing seed (the same key that
    is TOFU-pinned during the secure handshake), which binds the invite — and
    therefore the group roster snapshot — to the inviter's verified identity.
    """
    if not signing_seed:
        raise ValueError("A signing seed is required to sign a group invite")
    inviter_signing_pub = crypto.get_verify_key_from_seed(signing_seed).hex()
    created_at = _to_iso8601(invite.created_at or utc_now())
    expires_at = _to_iso8601(invite.expires_at) if invite.expires_at else None
    signed_bytes = _canonical_invite_bytes(
        version=GROUP_INVITE_VERSION,
        invite_id=invite.invite_id,
        group_id=invite.group_id,
        title=invite.title,
        members=invite.members,
        epoch=int(invite.epoch),
        inviter_id=invite.inviter_id,
        inviter_signing_pub=inviter_signing_pub,
        created_at=created_at,
        expires_at=expires_at,
    )
    signature = crypto.sign_data(signing_seed, signed_bytes).hex()
    payload = {
        "v": GROUP_INVITE_VERSION,
        "invite_id": invite.invite_id,
        "group_id": invite.group_id,
        "title": invite.title,
        "members": list(invite.members),
        "epoch": int(invite.epoch),
        "inviter_id": invite.inviter_id,
        "inviter_signing_pub": inviter_signing_pub,
        "created_at": created_at,
        "expires_at": expires_at,
        "signature": signature,
    }
    plaintext = json.dumps(payload, separators=(",", ":"), ensure_ascii=True).encode(
        "utf-8"
    )
    return _seal_invite_plaintext(plaintext)


def _derive_invite_wrap_key(wrap_key: bytes) -> bytes:
    prk = crypto.hkdf_extract(_INVITE_WRAP_SALT, wrap_key)
    return crypto.hkdf_expand(prk, _INVITE_WRAP_INFO, 32)


def _b64url_encode(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).decode("ascii").rstrip("=")


def _b64url_decode(text: str) -> bytes:
    compact = "".join((text or "").split())
    pad = "=" * ((4 - len(compact) % 4) % 4)
    try:
        return base64.urlsafe_b64decode(compact + pad)
    except (ValueError, TypeError) as exc:
        raise ValueError("Invalid group invite encoding") from exc


def _seal_invite_plaintext(plaintext: bytes) -> str:
    if not crypto.NACL_AVAILABLE:
        raise ValueError("PyNaCl is required to create a sealed group invite")
    wrap_key = secrets.token_bytes(_INVITE_WRAP_KEY_SIZE)
    file_key = _derive_invite_wrap_key(wrap_key)
    ciphertext = crypto.encrypt_message(file_key, plaintext)
    return _b64url_encode(wrap_key + ciphertext)


def _unseal_invite_plaintext(token: str) -> bytes:
    if not crypto.NACL_AVAILABLE:
        raise ValueError("PyNaCl is required to open a sealed group invite")
    compact = "".join((token or "").split())
    if not compact or any(ch not in _INVITE_B64URL_ALPHABET for ch in compact):
        raise ValueError("Not a group invite string")
    raw = _b64url_decode(compact)
    if len(raw) < _MIN_SEALED_INVITE_BYTES:
        raise ValueError("Group invite payload too small")
    wrap_key = raw[:_INVITE_WRAP_KEY_SIZE]
    ciphertext = raw[_INVITE_WRAP_KEY_SIZE:]
    plaintext = crypto.decrypt_message(_derive_invite_wrap_key(wrap_key), ciphertext)
    if plaintext is None:
        raise ValueError("Group invite decryption failed")
    return plaintext


def _load_invite_payload(text: str) -> dict[str, Any]:
    raw = (text or "").strip()
    if not raw:
        raise ValueError("Empty group invite payload")
    if len(raw) > MAX_GROUP_INVITE_BYTES:
        raise ValueError("Group invite payload too large")
    if raw.startswith(GROUP_INVITE_PREFIX):
        body = raw[len(GROUP_INVITE_PREFIX) :].strip()
        if not body:
            raise ValueError("Empty group invite payload")
        try:
            payload = json.loads(body)
        except json.JSONDecodeError as exc:
            raise ValueError("Invalid group invite JSON") from exc
    else:
        try:
            payload = json.loads(_unseal_invite_plaintext(raw).decode("utf-8"))
        except (UnicodeDecodeError, json.JSONDecodeError) as exc:
            raise ValueError("Invalid group invite payload") from exc
    if not isinstance(payload, dict):
        raise ValueError("Group invite payload must be an object")
    return payload


def decode_group_invite(text: str) -> GroupInvite:
    payload = _load_invite_payload(text)

    version = int(payload.get("v", 0))
    if version != GROUP_INVITE_VERSION:
        raise ValueError(
            f"Unsupported group invite version: {version} "
            f"(v1 unsigned invites are rejected)"
        )

    invite_id = str(payload.get("invite_id") or "").strip()
    group_id = str(payload.get("group_id") or "").strip()
    inviter_id = str(payload.get("inviter_id") or "").strip()
    title = str(payload.get("title") or "").strip() or None
    inviter_signing_pub = str(payload.get("inviter_signing_pub") or "").strip().lower()
    signature = str(payload.get("signature") or "").strip().lower()
    if not invite_id:
        raise ValueError("Missing required group invite field: invite_id")
    if not group_id:
        raise ValueError("Missing required group invite field: group_id")
    if not inviter_id:
        raise ValueError("Missing required group invite field: inviter_id")
    if len(inviter_signing_pub) != 64:
        raise ValueError("Missing or invalid inviter signing public key")
    if len(signature) != 128:
        raise ValueError("Missing or invalid group invite signature")

    try:
        epoch = int(payload.get("epoch", 0))
    except (TypeError, ValueError) as exc:
        raise ValueError("Invalid group invite epoch") from exc
    if epoch < 0:
        raise ValueError("Invalid group invite epoch")

    created_raw = payload.get("created_at")
    created_at: datetime | None = None
    if created_raw:
        try:
            created_at = _parse_datetime(str(created_raw))
        except Exception as exc:
            raise ValueError("Invalid group invite created_at") from exc

    expires_raw = payload.get("expires_at")
    expires_at: datetime | None = None
    if expires_raw:
        try:
            expires_at = _parse_datetime(str(expires_raw))
        except Exception as exc:
            raise ValueError("Invalid group invite expires_at") from exc

    members = _normalize_members(payload.get("members"))

    # Verify the inviter's signature over the canonical payload. This proves
    # the invite (and its member roster) was produced by the holder of
    # ``inviter_signing_pub`` and has not been tampered with in transit.
    try:
        signing_pub = bytes.fromhex(inviter_signing_pub)
        signature_bytes = bytes.fromhex(signature)
    except ValueError as exc:
        raise ValueError("Malformed group invite signature encoding") from exc
    normalized_inviter = normalize_member_id(inviter_id)
    signed_bytes = _canonical_invite_bytes(
        version=version,
        invite_id=invite_id,
        group_id=group_id,
        title=title,
        members=(
            members if normalized_inviter in members else members + (normalized_inviter,)
        ),
        epoch=epoch,
        inviter_id=normalized_inviter,
        inviter_signing_pub=inviter_signing_pub,
        # Use the raw payload strings so canonicalization matches the encoder
        # byte-for-byte (no datetime round-trip involved).
        created_at=str(created_raw or ""),
        expires_at=str(expires_raw) if expires_raw else None,
    )
    if not crypto.verify_signature(signing_pub, signed_bytes, signature_bytes):
        raise ValueError("Group invite signature verification failed")

    if expires_at is not None and expires_at < utc_now():
        raise ValueError("Group invite has expired")

    return GroupInvite(
        invite_id=invite_id,
        group_id=group_id,
        members=members,
        epoch=epoch,
        inviter_id=inviter_id,
        title=title,
        created_at=created_at,
        expires_at=expires_at,
        version=version,
        inviter_signing_pub=inviter_signing_pub,
        signature=signature,
    )


def looks_like_group_invite(text: str) -> bool:
    raw = (text or "").strip()
    if not raw:
        return False
    if raw.startswith(GROUP_INVITE_PREFIX):
        return True
    try:
        payload = _load_invite_payload(raw)
    except ValueError:
        return False
    return bool(payload.get("invite_id") and payload.get("signature"))

File: cpp/core/src/groups/invite.cpp
#include "i2pchat/groups/invite.hpp"

#include <algorithm>
#include <cctype>
#include <nlohmann/json.hpp>
#include <set>

#include "i2pchat/canonical_json.hpp"
#include "i2pchat/crypto.hpp"
#include "i2pchat/encoding.hpp"
#include "i2pchat/storage/chat_history.hpp"

namespace i2pchat::groups {
namespace {

constexpr std::string_view kWhitespace = " \t\r\n\f\v";
constexpr std::size_t kWrapKeySize = 32;
constexpr std::string_view kSealSalt = "I2PCHAT-GROUP-INVITE-SEAL";
constexpr std::string_view kSealInfo = "I2PCHAT-GROUP-INVITE-SEAL|v3";
/// wrap key + SecretBox overhead (24-byte nonce, 16-byte tag) + a minimal body.
constexpr std::size_t kMinSealedBytes = kWrapKeySize + 24 + 16 + 16;

std::string trim(std::string_view text) {
    const auto first = text.find_first_not_of(kWhitespace);
    if (first == std::string_view::npos) {
        return "";
    }
    const auto last = text.find_last_not_of(kWhitespace);
    return std::string(text.substr(first, last - first + 1));
}

std::string trim_lower(std::string_view text) {
    std::string value = trim(text);
    for (char& ch : value) {
        ch = static_cast<char>(std::tolower(static_cast<unsigned char>(ch)));
    }
    return value;
}

/// Whitespace anywhere is dropped: tokens get wrapped by mail clients and
/// pasted back with line breaks in them.
std::string strip_all_whitespace(std::string_view text) {
    std::string out;
    out.reserve(text.size());
    for (const char ch : text) {
        if (kWhitespace.find(ch) == std::string_view::npos) {
            out.push_back(ch);
        }
    }
    return out;
}

bool is_base64url_token(std::string_view text) {
    return !text.empty() && std::all_of(text.begin(), text.end(), [](const char ch) {
               return std::isalnum(static_cast<unsigned char>(ch)) != 0 || ch == '_' ||
                      ch == '-';
           });
}

std::vector<std::string> normalize_members(const nlohmann::json& raw) {
    if (!raw.is_array()) {
        throw WireError("Group invite members must be a list");
    }
    std::vector<std::string> members;
    std::set<std::string> seen;
    for (const nlohmann::json& item : raw) {
        const std::string text = item.is_string() ? item.get<std::string>() : item.dump();
        const std::string member = normalize_member_id(text);
        if (member.empty() || !seen.insert(member).second) {
            continue;
        }
        members.push_back(member);
    }
    if (members.empty()) {
        throw WireError("Group invite must include at least one member");
    }
    return members;
}

Bytes derive_seal_key(ByteView wrap_key) {
    const Bytes prk = crypto::hkdf_extract(as_bytes(kSealSalt), wrap_key);
    return crypto::hkdf_expand(ByteView(prk), as_bytes(kSealInfo), 32);
}

std::string seal(const std::string& plaintext) {
    const Bytes wrap_key = crypto::random_bytes(kWrapKeySize);
    const Bytes file_key = derive_seal_key(ByteView(wrap_key));
    const Bytes ciphertext = crypto::encrypt_message(ByteView(file_key), as_bytes(plaintext));

    Bytes blob;
    blob.reserve(wrap_key.size() + ciphertext.size());
    append(blob, ByteView(wrap_key));
    append(blob, ByteView(ciphertext));
    return encoding::base64url_encode_nopad(ByteView(blob));
}

std::string unseal(std::string_view token) {
    const std::string compact = strip_all_whitespace(token);
    if (!is_base64url_token(compact)) {
        throw WireError("Not a group invite string");
    }
    const std::optional<Bytes> raw = encoding::base64url_decode(compact);
    if (!raw.has_value()) {
        throw WireError("Invalid group invite encoding");
    }
    if (raw->size() < kMinSealedBytes) {
        throw WireError("Group invite payload too small");
    }
    const ByteView wrap_key = ByteView(*raw).subspan(0, kWrapKeySize);
    const ByteView ciphertext = ByteView(*raw).subspan(kWrapKeySize);
    const Bytes file_key = derive_seal_key(wrap_key);

    const std::optional<Bytes> plaintext =
        crypto::decrypt_message(ByteView(file_key), ciphertext);
    if (!plaintext.has_value()) {
        throw WireError("Group invite decryption failed");
    }
    return to_string(ByteView(*plaintext));
}

nlohmann::json load_payload(std::string_view text) {
    const std::string raw = trim(text);
    if (raw.empty()) {
        throw WireError("Empty group invite payload");
    }
    if (raw.size() > kMaxInviteBytes) {
        throw WireError("Group invite payload too large");
    }

    std::string body;
    if (raw.starts_with(kInvitePrefix)) {
        body = trim(std::string_view(raw).substr(kInvitePrefix.size()));
        if (body.empty()) {
            throw WireError("Empty group invite payload");
        }
    } else {
        body = unseal(raw);
    }

    nlohmann::json payload;
    try {
        payload = nlohmann::json::parse(body);
    } catch (const nlohmann::json::exception&) {
        throw WireError("Invalid group invite payload");
    }
    if (!payload.is_object()) {
        throw WireError("Group invite payload must be an object");
    }
    return payload;
}

std::string text_field(const nlohmann::json& payload, const char* key) {
    const auto it = payload.find(key);
    if (it == payload.end() || !it->is_string()) {
        return "";
    }
    return trim(it->get<std::string>());
}

/// The signed form. `expires_at` is `null` rather than absent when unset,
/// because the signature covers the field either way.
nlohmann::json signed_body(const GroupInvite& invite) {
    nlohmann::json payload = nlohmann::json::object();
    payload["created_at"] = invite.created_at;
    payload["epoch"] = invite.epoch;
    payload["expires_at"] =
        invite.expires_at.empty() ? nlohmann::json() : nlohmann::json(invite.expires_at);
    payload["group_id"] = invite.group_id;
    payload["invite_id"] = invite.invite_id;
    payload["inviter_id"] = invite.inviter_id;
    payload["inviter_signing_pub"] = invite.inviter_signing_pub;
    payload["members"] = invite.members;
    payload["title"] = invite.title.empty() ? nlohmann::json() : nlohmann::json(invite.title);
    payload["v"] = invite.version;
    return payload;
}

/// Members with the inviter appended when missing. The inviter is part of the
/// group by construction, and the signature is computed over this form.
std::vector<std::string> members_with_inviter(std::vector<std::string> members,
                                             const std::string& inviter_id) {
    if (std::find(members.begin(), members.end(), inviter_id) == members.end()) {
        members.push_back(inviter_id);
    }
    return members;
}

}  // namespace

Bytes invite_signature_payload(const GroupInvite& invite) {
    return to_bytes(std::string(kInviteSignatureDomain) + "|" +
                    json_canonical::dump(signed_body(invite)));
}

std::string encode_invite(const GroupInvite& invite, ByteView signing_seed) {
    if (signing_seed.empty()) {
        throw WireError("A signing seed is required to sign a group invite");
    }
    GroupInvite signed_invite = invite;
    signed_invite.invite_id = trim(invite.invite_id);
    signed_invite.group_id = trim(invite.group_id);
    signed_invite.inviter_id = normalize_member_id(invite.inviter_id);
    signed_invite.title = trim(invite.title);
    signed_invite.version = kInviteVersion;

    if (signed_invite.invite_id.empty()) {
        throw WireError("Group invite invite_id is required");
    }
    if (signed_invite.group_id.empty()) {
        throw WireError("Group invite group_id is required");
    }
    if (signed_invite.inviter_id.empty()) {
        throw WireError("Group invite inviter_id is required");
    }

    std::vector<std::string> members;
    std::set<std::string> seen;
    for (const std::string& raw : invite.members) {
        const std::string member = normalize_member_id(raw);
        if (member.empty() || !seen.insert(member).second) {
            continue;
        }
        members.push_back(member);
    }
    if (members.empty()) {
        throw WireError("Group invite must include at least one member");
    }
    signed_invite.members = members_with_inviter(std::move(members), signed_invite.inviter_id);

    signed_invite.inviter_signing_pub =
        encoding::hex_encode(ByteView(crypto::get_verify_key_from_seed(signing_seed)));
    signed_invite.signature = encoding::hex_encode(ByteView(crypto::sign_data(
        signing_seed, ByteView(invite_signature_payload(signed_invite)))));

    nlohmann::json payload = signed_body(signed_invite);
    payload["signature"] = signed_invite.signature;
    return seal(json_canonical::dump(payload));
}

GroupInvite decode_invite(std::string_view text,
                          std::optional<std::chrono::system_clock::time_point> now) {
    const nlohmann::json payload = load_payload(text);

    const auto version_field = payload.find("v");
    const int version = version_field != payload.end() && version_field->is_number_integer()
                            ? version_field->get<int>()
                            : 0;
    if (version != kInviteVersion) {
        throw WireError("Unsupported group invite version: " + std::to_string(version) +
                        " (v1 unsigned invites are rejected)");
    }

    GroupInvite invite;
    invite.version = version;
    invite.invite_id = text_field(payload, "invite_id");
    invite.group_id = text_field(payload, "group_id");
    invite.title = text_field(payload, "title");
    invite.created_at = text_field(payload, "created_at");
    invite.expires_at = text_field(payload, "expires_at");
    invite.inviter_signing_pub = trim_lower(text_field(payload, "inviter_signing_pub"));
    invite.signature = trim_lower(text_field(payload, "signature"));

    if (invite.invite_id.empty()) {
        throw WireError("Missing required group invite field: invite_id");
    }
    if (invite.group_id.empty()) {
        throw WireError("Missing required group invite field: group_id");
    }
    const std::string inviter_raw = text_field(payload, "inviter_id");
    if (inviter_raw.empty()) {
        throw WireError("Missing required group invite field: inviter_id");
    }
    invite.inviter_id = normalize_member_id(inviter_raw);
    if (invite.inviter_id.empty()) {
        throw WireError("Missing required group invite field: inviter_id");
    }
    if (invite.inviter_signing_pub.size() != 64) {
        throw WireError("Missing or invalid inviter signing public key");
    }
    if (invite.signature.size() != 128) {
        throw WireError("Missing or invalid group invite signature");
    }

    const auto epoch_field = payload.find("epoch");
    if (epoch_field == payload.end() || !epoch_field->is_number_integer() ||
        epoch_field->get<std::int64_t>() < 0) {
        throw WireError("Invalid group invite epoch");
    }
    invite.epoch = epoch_field->get<std::uint64_t>();

    if (!invite.created_at.empty() &&
        !storage::parse_iso8601_utc(invite.created_at).has_value()) {
        throw WireError("Invalid group invite created_at");
    }
    std::optional<std::chrono::system_clock::time_point> expires;
    if (!invite.expires_at.empty()) {
        expires = storage::parse_iso8601_utc(invite.expires_at);
        if (!expires.has_value()) {
            throw WireError("Invalid group invite expires_at");
        }
    }

    const auto members_field = payload.find("members");
    invite.members = normalize_members(members_field == payload.end() ? nlohmann::json()
                                                                     : *members_field);

    const std::optional<Bytes> signing_pub = encoding::hex_decode(invite.inviter_signing_pub);
    const std::optional<Bytes> signature = encoding::hex_decode(invite.signature);
    if (!signing_pub.has_value() || !signature.has_value()) {
        throw WireError("Malformed group invite signature encoding");
    }

    // Verify over the roster as the encoder signed it, with the inviter present,
    // rather than over what the payload happens to list. That is what proves the
    // roster itself was not edited in transit.
    GroupInvite signed_form = invite;
    signed_form.members = members_with_inviter(invite.members, invite.inviter_id);
    if (!crypto::verify_signature(ByteView(*signing_pub),
                                  ByteView(invite_signature_payload(signed_form)),
                                  ByteView(*signature))) {
        throw WireError("Group invite signature verification failed");
    }

    if (expires.has_value() && *expires < now.value_or(std::chrono::system_clock::now())) {
        throw WireError("Group invite has expired");
    }

    invite.members = std::move(signed_form.members);
    return invite;
}

bool looks_like_invite(std::string_view text) {
    const std::string raw = trim(text);
    if (raw.empty()) {
        return false;
    }
    if (raw.starts_with(kInvitePrefix)) {
        return true;
    }
    try {
        const nlohmann::json payload = load_payload(raw);
        return !text_field(payload, "invite_id").empty() &&
               !text_field(payload, "signature").empty();
    } catch (const std::exception&) {
        return false;
    }
}

}  // namespace i2pchat::groups
